### ISSUE TITLE
Pass numbers as strings from Perl to JavaScript

### ISSUE THREAD

**Issue by bschmalhofer** [RotherOSS/otobo#2595]
It was already stated in the parent issue, #2371  that the problem can be highlighted by dumping a numeric Sysconfig setting one time with JSON and the other time Cpanel::JSON::XS.

otobo@ad98e2ba4b0f:~$ cp JSON_old_but_working.pm Kernel/System/JSON.pm 
otobo@ad98eis2ba4b0f:~$ bin/otobo.Console.pl Maint::Config::Dump SessionUseCookie
"0"
otobo@ad98e2ba4b0f:~$ cp JSON_new_but_broken.pm Kernel/System/JSON.pm 
otobo@ad98e2ba4b0f:~$ bin/otobo.Console.pl Maint::Config::Dump SessionUseCookie
0
otobo@ad98e2ba4b0f:~$ 

Further debugging showed that the versions of the JSON libraries do not matter.

The explanation why SessionUseCookie is considered one time as a string and the other time as number is not obvious. But the relevant hint is right in the documentation of Cpanel::JSON::XS: https://metacpan.org/pod/Cpanel::JSON::XS#simple-scalars .

    Non-deterministic behavior is following: scalars that have last been used in a string context before encoding as JSON strings, and anything else as number value:

For SessionUseCookie this checks out. The number is created by running ZZZAAuto,pm and never meddled with, e.g. printed, before converted to JSON.

JSON::XS follows more or less the same strategy but apparently there are differences in details.

The take home message is that the decision why a (numeric) value is represented as JSON as a string might be deterministic, but not transparent. So the next best idea is to force transparency. This can be done by setting $json->type_all_string(1) for the values that are passed into the JavaScript World. This is a simple rule, that forces stringiness on the numbers. It seems to be in line with the current implementation:

otobo$ grep -r Core.Config.Get  var/httpd/htdocs/js/ | grep parseInt | wc -l
27

_Originally posted by @bschmalhofer in https://github.com/RotherOSS/otobo/issues/2371#issuecomment-1777574124_

**Comment by bschmalhofer** [comment]
Talked about this with @svenoe  and we concluded that we should try the "type_all_string" approach.

**Comment by bschmalhofer** [comment]
Simply setting type_all_string did the trick, but with a bad twist. The boolean values are now also encoded as strings, with the caveat that the string `"false"` happens to be true. See also https://github.com/rurban/Cpanel-JSON-XS/issues/175.

This behavior can be seen with the following sample code:

```
otobo$ cat t.pl 
use strict;
use warnings;
use feature qw(say);

use Cpanel::JSON::XS;
use Kernel::System::JSON;

my $JSONObject = Kernel::System::JSON->new;

my $json = Cpanel::JSON::XS->new->allow_nonref;

sub say_something {
    my ($msg) = @_;

    say $msg;
    say '';

    say 'true values:';
    say $json->encode(\1);
    say $json->encode(Cpanel::JSON::XS::true);
    say $json->encode($JSONObject->True);

    say '';

    say 'false values:';
    say $json->encode(\0);
    say $json->encode(Cpanel::JSON::XS::false);
    say $json->encode($JSONObject->False);

    say '';
}

say_something('no type_all_string');

$json->type_all_string;

say_something('type_all_string');


otobo$ perl -I . t.pl 
no type_all_string

true values:
true
true
true

false values:
false
false
false

type_all_string

true values:
"SCALAR(0x5589227e2fa8)"
"true"
"true"

false values:
"SCALAR(0x558922937958)"
"false"
"false"


```

This is bad, but not very bad. Boolean values, that is true and false, are not really needed for getting the relevant information from Perl to JavaScript. Simply use `"1"` and `""` for that. For convenience the method `AddJSBoolean()`is provided which emits these workaround strings.

**Comment by bschmalhofer** [comment]
The test suite looks mostly fine. The failure in DynamicFieldDatabase.t  seems to be unrelated. The failures in Render.t and JSON.t have already been taken care of. The failure in JSDataAgent.t still needs to be investigated.                                            

> 
> Test Summary Report
> -------------------
> /opt/otobo/scripts/test/Compile.t                                                                         (Wstat: 0 Tests: 1724 Failed: 0)
>   TODO passed:   687, 812
> /opt/otobo/scripts/test/JSON.t                                                                            (Wstat: 6400 (exited 25) Tests: 0 Failed: 0)
>   Non-zero exit status: 25
>   Parse errors: No plan found in TAP output
> /opt/otobo/scripts/test/Layout/Template/Render.t                                                          (Wstat: 256 (exited 1) Tests: 41 Failed: 1)
>   Failed test:  30
>   Non-zero exit status: 1
> /opt/otobo/scripts/test/Selenium/Customer/DynamicFields/DynamicFieldDatabase.t                            (Wstat: 256 (exited 1) Tests: 137 Failed: 1)
>   Failed test:  137
>   Non-zero exit status: 1
> /opt/otobo/scripts/test/Selenium/JavaScript/JSDataAgent.t                                                 (Wstat: 256 (exited 1) Tests: 28 Failed: 1)
>   Failed test:  9
>   Non-zero exit status: 1
> /opt/otobo/scripts/test/Selenium/TestingMethods.t                                                         (Wstat: 0 Tests: 36 Failed: 0)
>   TODO passed:   21, 24, 28
> Files=961, Tests=80652, 4211 wallclock secs (15.41 usr  2.36 sys + 1198.81 cusr 98.56 csys = 1315.14 CPU)
> Result: FAIL
> ran tests for product OTOBO 11.0.x on host e10a896b176c .
>

**Comment by bschmalhofer** [comment]
Discussed this with @svenoe . The attribute type_all_string of `Cpanel::JSON::JS`can't be used, as the minimum version for that feature is not easily available on the supported distributions. So, we reverted back to `JSON::XS`. See also https://github.com/RotherOSS/otobo/pull/2600 .

The test suite looks fine. Closing this issue.

Edit: the rollback was done in https://github.com/RotherOSS/otobo/commit/7a87c0c2f00403a1921ee4f1491578740632d594 . Sorry for the incorrect issue number.
